apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,921 @@
+      () => executeNavigationStep('beforeLeave')
+    ).then(
+      () => window.history.back()
+    ).then(
+      () => navigationTransition.defer.promise,
+      onNavigationError
+    );
+    // todo reset navigation on "pageshow" event ?
   },
 
-  getPanelName: function n_getPanelName() {
-    var hash = window.location.hash;
-    if (hash.length && hash.startsWith('#')) {
-      hash = hash.slice(1);
+  /**
+   * Lifecycle methods are generally called in this order:
+   * - previousPanel.beforeLeave
+   * - nextPanel.beforeEnter
+   * - nextPanel.afterEnter
+   * - previousPanel.afterLeave
+   *
+   * If we stay in the same document and enter a new panel, the panels are slid
+   * between beforeEnter and afterLeave.
+   *
+   * With the new architecture, the navigation can happen in one of 2 cases:
+   * 1. either we stay in the same document,
+   * 2. or we move to another document.
+   *
+   * In all cases:
+   * - beforeLeave can prevent changing the location and thus the
+   *   navigation. It is expected that beforeLeave does not change any state but
+   *   merely checks whether the transition can take place.
+   * - beforeEnter will be called before the view is displayed. Therefore it can
+   *   make visual changes but it can't prevent a view from being entered. All
+   *   rejections and exceptions will be swallowed and output to the console.
+   * - afterEnter does left-over actions
+   * - afterLeave needs to clean up the panel, reclaim memory. It's called only
+   *   when the user stays in the same document.
+   *
+   * toPanel itself returns a Promise that will be resolved if the transition
+   * happened and rejected otherwise.
+   *
+   * @param {String} viewName The view to load.
+   * @param {Object} [args] The arguments to pass to the view.
+   * @returns {Promise} Resolves when the transition is over.
+   */
+  toPanel(viewName, args) {
+    debug('toPanel(%s)', viewName);
+    var view = findViewFromName(viewName);
+
+    if (!view) {
+      return Promise.reject(
+        new Error(`Couldn't find view ${viewName}`)
+      );
+    }
+
+    if (this.isCurrentPanel(viewName, args)) {
+      debug('View %s is the current view.', viewName);
+      return Promise.resolve();
     }
 
-    var index;
-    if ((index = hash.indexOf('&')) !== -1) {
-      hash = hash.slice(0, index);
+    var hash = '#';
+    if (view.partOf) {
+      hash += '/' + viewName;
     }
+    hash = Utils.url(hash, args);
+
+    return startNavigation(
+      { newPanel: view, args, partial: true }
+    ).then(
+      () => executeNavigationStep('beforeLeave')
+    ).then(() => {
+      var nextLocation;
+      if (navigationTransition.leaveDocument) {
+        var mainView = findMainView(view);
+        nextLocation = '/' + mainView.url + hash;
+      } else {
+        nextLocation = hash;
+      }
 
-    return hash;
+      setLocation(nextLocation);
+    }).then(
+      () => navigationTransition.defer.promise,
```

Comment:
__question:__ mm can we use ```endNavigation``` here as well? ```endNavigation``` is the only place that fires ```navigated``` event.